File: supabase/functions/generate-summary/topic_summary_processor.ts
import { ContentType } from "../_shared/base_processor.ts";
import { BaseSummaryProcessor, Summary, SummaryContent } from "./base_summary_processor.ts";
export class TopicSummaryProcessor extends BaseSummaryProcessor {
    private topics: SummaryContent;
    private topicNames: string[];
    constructor(
        apiKey: string,
        courseTitle: string,
        topicNames: string[],
        topics: SummaryContent,
    ) {
        super(apiKey, courseTitle, ContentType.TOPIC);
        this.topics = topics;
        this.topicNames = topicNames;
    }

    async processSummary(
        allLectures: {note_number: number, id: string}[], 
        numBatches: number, 
        onBatchComplete: (batchNumber: number, summary: Summary) => Promise<void>
    ): Promise<Summary> {
        console.log(`Generating summary for ${this.topicNames}`);
        
        const names = this.topicNames.join(", ");
        const batches = this.splitContentIntoBatches(this.topics.content, numBatches);
        
        for (let i = 0; i < batches.length; i++) {
            console.log(`Processing batch ${i + 1} of ${batches.length}`);
            const result = await this.processBatch(
                names,
                batches[i],
            );
            console.log(`Batch ${i + 1} result:`, result);
            this.cleanResult(result, names, allLectures);
            
            // Call the batch completion callback
            await onBatchComplete(i + 1, this.summary[names]);
        }

        return this.summary[names];
    }
}

File: server/app/services/summary/topic_summary_processor.py
from typing import List, Dict, Callable, Awaitable
from app.services.base_processor import ContentType
from app.services.summary.base_summary_processor import BaseSummaryProcessor, Summary, SummaryContent

class TopicSummaryProcessor(BaseSummaryProcessor):
    def __init__(
        self,
        course_title: str,
        topic_names: List[str],
        topics: SummaryContent,
        additional_instructions: str
    ):
        super().__init__(course_title, ContentType.TOPIC, additional_instructions)
        self.topics = topics
        self.topic_names = topic_names
    async def process_summary(
        self,
        all_lectures: List[Dict[str, str]],
        num_batches: int,
        on_batch_complete: Callable[[int, Summary], Awaitable[None]]
    ) -> Summary:
        """
        Process the topic content in batches and generate summaries.
        
        Args:
            all_lectures: List of lecture dictionaries with note_number and id
            num_batches: Number of batches to split the content into
            on_batch_complete: Callback function to execute after each batch
        """
        print(f"Generating summary for {self.topic_names}")
        
        names = ", ".join(self.topic_names)
        batches = self.split_content_into_batches(self.topics['content'], num_batches)
        
        for i, batch in enumerate(batches):
            print(f"Processing batch {i + 1} of {len(batches)}")
            result = await self.process_batch(names, batch)
            print(f"Batch {i + 1} result:", result)
            self.clean_result(result, names, all_lectures)
            
            # Call the batch completion callback
            await on_batch_complete(i + 1, self.summary[names])

        return self.summary[names] 